scipy/stats/_distn_infrastructure.py

```diff
@@ -554,18 +555,20 @@ def _parse_args_stats(self, %(shape_arg_str)s %(locscale_in)s, moments='mv'):
 # The function name ncx2 is an abbreviation for noncentral chi squared.
 
 def _ncx2_log_pdf(x, df, nc):
+    pdf = _ncx2_pdf(x, df, nc)
+    return _lazywhere(pdf > 0, (pdf,), f=np.log, fillvalue=-np.inf)
+
+
+def _ncx2_pdf(x, df, nc):
+    # Calculate logpdf then take exp.
     # We use (xs**2 + ns**2)/2 = (xs - ns)**2/2  + xs*ns, and include the
     # factor of exp(-xs*ns) into the ive function to improve numerical
     # stability at large values of xs. See also `rice.pdf`.
     df2 = df/2.0 - 1.0
     xs, ns = np.sqrt(x), np.sqrt(nc)
     res = xlogy(df2/2.0, x/nc) - 0.5*(xs - ns)**2
-    res += np.log(ive(df2, xs*ns) / 2.0)
-    return res
-
-
-def _ncx2_pdf(x, df, nc):
-    return np.exp(_ncx2_log_pdf(x, df, nc))
+    corr = ive(df2, xs*ns) / 2.0
+    return np.exp(res) * corr
```

Comment:
In the `pdf` case this is an improvement. 
old code: `np.exp(res + np.log(ive))`
new code: `np.exp(res) * ive`

for the `log_pdf` case this is a regression
old code: `res + np.log(ive)`
new code: `_lazywhere(np.log(np.exp(res) * ive))`